File: medium/3532_path_existence_queries_in_a_graph_i/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<bool>
  pathExistenceQueries(int n, std::vector<int> &nums, int maxDiff,
                       std::vector<std::vector<int>> &queries) {
    std::vector<int> groupsMap(n);
    int currGroup = 0;

    for (size_t i = 1; i < n; ++i) {
      if (nums[i] - nums[i - 1] > maxDiff)
        ++currGroup;

      groupsMap[i] = currGroup;
    }

    std::vector<bool> pathsExist;
    for (const std::vector<int> &query : queries)
      pathsExist.push_back(groupsMap[query[0]] == groupsMap[query[1]]);

    return pathsExist;
  }
};

File: medium/3532_path_existence_queries_in_a_graph_i/solution.py
from typing import List


class Solution:
    def pathExistenceQueries(
        self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]
    ) -> List[bool]:
        groupsMap = [0] * n
        currGroup = 0

        for i in range(1, n):
            if nums[i] - nums[i - 1] > maxDiff:
                currGroup += 1

            groupsMap[i] = currGroup

        return [groupsMap[i] == groupsMap[j] for i, j in queries]
